File: problem64.cpp
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        // This is DP problem
        // Need n and m of grid 
        int n = grid.size();
        int m = grid[0].size();

        // Change all numbers in first row and col 
        for(int i = 1; i < n; i++){
            grid[i][0] += grid[i-1][0];
        }

        for(int j =1; j < m; j++){
            grid[0][j] += grid[0][j-1];
        }

        for(int row = 1; row < n; row++){
            for(int col = 1; col < m; col++){
                grid[row][col] += min(grid[row-1][col], grid[row][col-1]);
            }
        }

        return grid[n-1][m-1];
    }
};

File: problem64.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        n, m = len(grid), len(grid[0])

        # Fill the first row (can only come from the left)
        for col in range(1, m):
            grid[0][col] += grid[0][col - 1]

        # Fill the first column (can only come from above)
        for row in range(1, n):
            grid[row][0] += grid[row - 1][0]

        # Fill the rest of the grid
        for row in range(1, n):
            for col in range(1, m):
                grid[row][col] += min(grid[row - 1][col], grid[row][col - 1])

        return grid[n - 1][m - 1]